contracts/vortex/CarbonVortex.sol

```diff
@@ -0,0 +1,259 @@
+     */
+    function __CarbonVortex_init_unchained() internal onlyInitializing {
+        setRewardsPPM(100_000);
+    }
+
+    /**
+     * @dev authorize the contract to receive the native token
+     */
+    receive() external payable {}
+
+    /**
+     * @inheritdoc Upgradeable
+     */
+    function version() public pure override(IVersioned, Upgradeable) returns (uint16) {
+        return 2;
+    }
+
+    /**
+     * @inheritdoc ICarbonVortex
+     */
+    function setRewardsPPM(uint256 newRewardsPPM) public onlyAdmin validFee(uint32(newRewardsPPM)) {
+        uint256 prevRewardsPPM = _rewardsPPM;
+
+        // return if the rewards percentage PPM is the same
+        if (prevRewardsPPM == newRewardsPPM) {
+            return;
+        }
+
+        _rewardsPPM = newRewardsPPM;
+
+        emit RewardsUpdated({ prevRewardsPPM: prevRewardsPPM, newRewardsPPM: newRewardsPPM });
+    }
+
+    /**
+     * @inheritdoc ICarbonVortex
+     */
+    function rewardsPPM() external view returns (uint256) {
+        return _rewardsPPM;
+    }
+
+    /**
+     * @inheritdoc ICarbonVortex
+     */
+    function totalBurned() external view returns (uint256) {
+        return _totalBurned;
+    }
+
+    /**
+     * @inheritdoc ICarbonVortex
+     */
+    function availableFees(Token token) external view returns (uint256) {
+        return _carbonController.accumulatedFees(token) + token.balanceOf(address(this));
+    }
+
+    /**
+     * @inheritdoc ICarbonVortex
+     */
+    function execute(Token[] calldata tokens) external nonReentrant validateTokens(tokens) {
+        uint256 len = tokens.length;
+
+        // allocate balances array for the tokens
+        uint256[] memory balances = new uint256[](len);
+        // allocate array for the reward amounts for caller
+        uint256[] memory rewardAmounts = new uint256[](len);
+        // cache rewardsPPM to save gas
+        uint256 rewardsPercentage = _rewardsPPM;
+
+        // withdraw fees, load balances and reward amounts
+        for (uint256 i = 0; i < len; i = uncheckedInc(i)) {
+            // withdraw token fees
+            _carbonController.withdrawFees(tokens[i], type(uint256).max, address(this));
+            // get token balance
+            balances[i] = tokens[i].balanceOf(address(this));
+            // get reward amount for token
+            rewardAmounts[i] = MathEx.mulDivF(balances[i], rewardsPercentage, PPM_RESOLUTION);
+        }
+
+        // convert tokens to BNT
+        for (uint256 i = 0; i < len; i = uncheckedInc(i)) {
+            Token token = tokens[i];
+            // skip token if no token balance or token is BNT - no need to swap in this case
+            if (balances[i] == 0 || token == _bnt) {
+                continue;
+            }
+
+            // get trade amount for token
+            uint256 tradeAmount = balances[i] - rewardAmounts[i];
+
+            // approve tokens for trading on Bancor Network V3
+            _setAllowance(token, tradeAmount);
```

Comment:
Instead of increasing the allowance, which is currently resulting in two `allowance` external calls (one here and one in `SafeERC20`'s `safeIncreaseAllowance`) and an `approve` external call you should consider (mutually exclusive):

1. Remove the first redundant `allowance` call
2. Set the allowance to `type(uint256).max` only once by either:
a. Have a global (not even permissioned) `approveBancorNetwork` function which will approve `_bancorNetwork` for a token/list of tokens.
b. Have a mapping of which tokens were already approved and check and set it once (it's less gas efficient than the previous option, but removes the need for a global function to handle this).

BTW, please note that the approach won't work for tokens (such as USDT) which require setting the approval first to 0, in case that for some reason the approval wasn't exact.

Reply:
Actually, we do two calls only on the first approval for a token - after it will be only the `allowance` call.
`uint256` is a huge number and for most tokens, during the lifetime of the contract, that will be enough.
So the only extra call will be to `allowance` for each token. 
While I agree it will save more gas to have a centralized function to approve the tokens, there may be new tokens in the future not supported on V3, and there may be more changes, so this approach is most flexible at the overhead of one external call.

Reply:
IMHO, managing allowances externally would be still better especially to the reasons that you've mentioned (would need to support other tokens, other approval targets, etc.), but it can be done later.

Another (more complicated) alternative is to actually try to do the trade without an approval and `try/catch` the revert from V3 and only then `approve` and try again. This will be more complicated to code, but will also result in no `allowance` calls in a happy flow.

Reply:
I think it could be worth it to do the external allowance management at a later point. 
In terms of gas, it will save 2100 gas per token in the call. Still, it will require tracking of the tokens which are going to be supported and manually approving them, which is a bit of operational overhead.

The second approach is rather interesting, I could consider implementing it.
It'd be good to hear some other opinions on this.

Reply:
I’d keep it simple